Mathlib/Tactic/Linter/Style.lean

```diff
@@ -0,0 +1,69 @@
+/-
+Copyright (c) 2024 Michael Rothgang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Rothgang
+-/
+
+import Lean.Elab.Command
+import Lean.Linter.Util
+
+/-!
+## Style linters
+
+This file contains (currently one, eventually more) linters about stylistic aspects:
+these are only about coding style, but do not affect correctness nor global coherence of mathlib.
+Historically, these were ported from the `lint-style.py` Python script.
+-/
+
+open Lean Elab Command
+
+namespace Mathlib.Linter.Style.SetOption
+
+/-- Whether a syntax element is a `set_option` call:
+Return the name of the option being set, if any. -/
+def parse_set_option : Syntax → Option (Ident)
+  -- This handles all four possibilities of `_val`: a string, number, `true` and `false`
+  | `(command|set_option $name:ident $_val) => some name
+  | `(set_option $name:ident $_val in $_x) => some name
+  | `(tactic|set_option $name:ident $_val in $_x) => some name
+  | _ => none
+
+/-- Whether a given piece of syntax is a `set_option` command, tactic or term. -/
+def is_set_option : Syntax → Bool :=
+  fun stx ↦ parse_set_option stx matches some _name
+
+/-- The `setOption` linter emits a warning on a `set_option` command, term or tactic
+which sets a `pp`, `profiler` or `trace` option. -/
+register_option linter.setOption : Bool := {
+  defValue := true
+  descr := "enable the `setOption` linter"
+}
+
+/-- Gets the value of the `linter.setOption` option. -/
+def getLinterHash (o : Options) : Bool := Linter.getLinterValue linter.setOption o
+
+/-- The `setOption` linter: this lints any `set_option` command, term or tactic
+which sets a `pp`, `profiler` or `trace` option.
+
+**Why is this bad?** These options are good for debugging, but should not be
+used in production code.
+**How to fix this?** Remove these options: usually, they are not necessary for production code.
+(Some tests will intentionally use one of these options; in this case, simply allow the linter.)
+-/
+def setOptionLinter : Linter where run := withSetOptionIn fun stx => do
+    unless getLinterHash (← getOptions) do
+      return
+    if (← MonadState.get).messages.hasErrors then
+      return
+    match stx.findStack? (fun _ ↦ true) is_set_option with
```

Comment:
It seems like `findStack?` isn't an ideal API for this use case. I would think `Syntax` should provide a function like `find? (root : Syntax) (needle : Syntax → Option β) : Option β` or `List β`. This would avoid having to `parse_set_option` twice. Overall, probably not worth bothering with unless this pattern comes up a lot when linting.

Reply:
Good point. I've switched to Syntax.find? now - this is not quite the API you described, but closer (and more concise also). It still requires parsing the `set_option` call twice, but if there are no errors, this hopefully doesn't matter.